### PR TITLE
[WIP/RFC] Windows TUI

### PR DESCRIPTION
**Pull request by equalsraf**
Enable the TUI for Windows. libuv's tty is used to get the console input, which is then passed to libtermkey. I've included a commit to enable the TUI, but we can also delay that for another PR

Main issues:

- [ ] #6659 libtermkey is being downloaded from my repo, for the windows fixes
- [x] the cursor does not seem right, it always looks the same - I think libuv does not support cursor shapes, so we would need to address it upstream
- [ ] #6660 :stop is broken, there is no implementation for suspend
- [ ] #6661 closing the window in the close button displays a message about SIGHUP, takes a while to shut down too - is there a reason for SIGHUP shutdown to take a while?
- [x] Loading a colorscheme from init.vim does not produce the same outcome as doing later on - see https://github.com/neovim/neovim/pull/6315#issuecomment-287657577
- [ ] Some text colors/attributes do not look right, it does look consistent so maybe were are sending a sequence we should not, because the terminal does not support it
- [ ] The Home/End keys do not work
- [ ] Keys F1, F2, F3, F4, F5 input A,B,C,D respectively but F6-F12 work fine. pressing `<C-V> + F1` also prints A
- [ ] terminal content scrolling seems broken in some cases. e.g. `nvim -u NONE` works but with color=wombat256mod the terminal keeps pushing data up

@justinmk I suspect it will be hard for me to follow up with this for the next couple weeks. If anyone wants to take over this, they have my blessings. Whoever can provide some testing time it would be much appreciated too.

### References

- [libuv's win/tty.c](https://github.com/libuv/libuv/blob/v1.x/src/win/tty.c)

### CODE REVIEW

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -775,26 +802,6 @@ static void unibi_goto(UI *ui, int row, int col)
   unibi_out(ui, unibi_cursor_address);
 }
 
-static void unibi_out(UI *ui, int unibi_index)
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: Why move this function?

**Comment by justinmk** [comment]
Thanks @equalsraf. Ideally we could upstream the libtermkey changes, is https://github.com/equalsraf/libtermkey/commit/b9317d6473b7a598359ebdfae5fd228bc159d0de the only commit needed? cc @leonerd

**Comment by equalsraf** [comment]
> is equalsraf/libtermkey@b9317d6 the only commit needed?

Yes.

File: src/nvim/tui/tui.c (hunk 1 above)
Comment (equalsraf) [inline src/nvim/tui/tui.c]: No reason, likely a result of rebasing a very old commit. I've cleaned up the patch.

**Comment by equalsraf** [comment]
(Windows 8) Some issues with colors schemes. Two instances of nvim with the color scheme wombat256. The right one loaded the colorscheme from init.vim. In the left one it was loaded manually after starting. Some other issues also affect scrolling on the right one too:

term

**Comment by justinmk** [comment]
Tried the appveyor artifact from this PR. Definitely some rendering issues, but a great start. Enough to get some basic test coverage going.

**Comment by equalsraf** [comment]
Have not verified, but someone just reported that Home/End keys are not working at all (Windows 10, on ConEmu).

**Comment by janlazo** [comment]
How do I stop the buffer from scrolling up for `:!` commands? I have to use `redraw!` or similar to reset the buffer.

**Comment by justinmk** [comment]
@janlazo I think that's just a quirk we haven't fixed yet. This PR is just getting the ball rolling, help welcome.

**Comment by justinmk** [comment]
Restarted the travis build, the failures were due to a cached version of the lua-client.

Let's merge this, it will help with troubleshooting bug reports on Windows by eliminating the GUI.

**Comment by equalsraf** [comment]
@justinmk maybe move the libtermkey stuff into neovim/libtermkey first

**Comment by justinmk** [comment]
Merged. @equalsraf Well, we're downloading a binary so it doesn't really matter. We can fix it up in https://github.com/neovim/neovim/issues/6659

**Pull request closed without merging**